src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServer/Handler/Restore/RestoreHandler.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+
+/// <summary>
+/// Given an input project (or none), runs restore on the project and streams the output
+/// back to the client to display.
+/// </summary>
+[ExportCSharpVisualBasicStatelessLspService(typeof(RestoreHandler)), Shared]
+[Method(MethodName)]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class RestoreHandler(DotnetCliHelper dotnetCliHelper) : ILspServiceRequestHandler<RestoreParams, RestorePartialResult[]>
+{
+    internal const string MethodName = "workspace/_roslyn_restore";
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public async Task<RestorePartialResult[]> HandleRequestAsync(RestoreParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(context.Solution);
+        using var progress = BufferedProgress.Create(request.PartialResultToken);
+
+        progress.Report(new RestorePartialResult(LanguageServerResources.Restore, LanguageServerResources.Restore_started));
+
+        var restorePaths = GetRestorePaths(request, context.Solution, context);
+        if (restorePaths.IsEmpty)
+        {
+            progress.Report(new RestorePartialResult(LanguageServerResources.Restore, LanguageServerResources.Nothing_found_to_restore));
+            return progress.GetValues() ?? [];
+        }
+
+        await RestoreAsync(restorePaths, progress, cancellationToken);
+
+        progress.Report(new RestorePartialResult(LanguageServerResources.Restore, $"{LanguageServerResources.Restore_complete}{Environment.NewLine}"));
+        return progress.GetValues() ?? [];
+    }
+
+    private async Task RestoreAsync(ImmutableArray<string> pathsToRestore, BufferedProgress<RestorePartialResult> progress, CancellationToken cancellationToken)
+    {
+        foreach (var path in pathsToRestore)
+        {
+            var arguments = $"restore \"{path}\"";
```

Comment:
Does this need other escaping if the file path contains a " or something? .NET Core lets you pass around argument lists directly to ProcessStartInfo so I'd feel better doing that if we could.